doc/classes/Dictionary.xml

```diff
@@ -105,7 +105,8 @@
 			<argument index="0" name="key" type="Variant">
 			</argument>
 			<description>
-				Erase a dictionary key/value pair by key. Returns [code]true[/code] if the given key was present in the dictionary, [code]false[/code] otherwise. Does not erase elements while iterating over the dictionary.
+				Erase a dictionary key/value pair by key. Returns [code]true[/code] if the given key was present in the dictionary, [code]false[/code] otherwise.
+				[b]Note:[/b] Do not erase elements while iterating over the dictionary.
```

Comment:
It might be useful to explain what could happen if you erase elements while iterating over the dictionary. In other words, people need to know *why* they should avoid doing this :slightly_smiling_face:

Reply:
Yes, I agree. However, that's programming 101 and also applicable to arrays, lists, and binary trees etc. Not sure a paragraph explaining it in depth is suitable here.

Reply:
A lot of people are requesting more examples in the documentation, so I'd think it'd be worth mentioning even if it can be obvious to more experienced programmers.

Reply:
I wouldn't make it too verbose and in-depth, but mentioning what happens if you do attempt to remove elements while iterating would be good (especially, does it fail, does it crash, does it lead to UB?).

Reply:
> Yes, I agree. However, that's programming 101 and also applicable to arrays, lists, and binary trees etc. Not sure a paragraph explaining it in depth is suitable here.

In most languages it's completely fine to erase elements during iteration, and if there are some considerations to take into account, they should be detailed.

However, since there is no way to directly access and update iterator in GDScript's `for x in dict:` loop, it's not practical to do so:

- In case of `Dictionary` `x` is a key and used as iterator, Godot using following code to advance it:
```c++
	OrderedHashMap<....>::Element E = _p->variant_map.find(*p_key);
	if (E && E.next()) {
		return &E.next().key();
	}
	return nullptr;
```
If the current element is deleted, it won't find it on next iteration and is going to exit the loop.

- In case of `Array`, `x` is value, and there's internal index counter, boundary checks are done on each iteration. If element at the current position or before it is erased, some elements will be skipped.

Reply:
And probably it's better to provide useful example, something like:
"If you want to erase Dictionary elements use `for a in dict.keys():` loop instead of `for a in dict:`."

Reply:
I was going to propose the text below, but according to @bruvzg, the behaviour is defined. Is that behaviour by design, or did it just happen to be? Do you want to publicly document this behaviour?

Honestly, this feels like something the core devs should decide on and not something I can comment on, other than that it's not a behaviour that I'd expect. Either it iterates over all items as set when the iteration began, nicely handling any erased items, or erasing during iteration is simply not supported.

> The underlying problem is that the iterators, i.e. the "pointers" into the container, are not aware of that the contents of the container has changed when erase() is called which leads to these iterators pointing to invalid data/memory areas which in turn leads to either undefined behaviour or, more often, a crash.